rust/src/delta_datafusion.rs

```diff
@@ -603,11 +606,84 @@ fn left_larger_than_right(left: ScalarValue, right: ScalarValue) -> Option<bool>
     }
 }
 
+/// Responsible for checking batches of data conform to table's invariants.
+pub struct DeltaDataChecker {
+    invariants: Vec<Invariant>,
+    ctx: SessionContext,
+    rt: tokio::runtime::Runtime,
+}
+
+impl DeltaDataChecker {
+    /// Create a new DeltaDataChecker
+    pub fn new(invariants: Vec<Invariant>) -> Self {
+        Self {
+            invariants,
+            ctx: SessionContext::new(),
+            rt: tokio::runtime::Runtime::new().unwrap(),
+        }
+    }
+
+    /// Check that a record batch conforms to table's invariants.
+    ///
+    /// If it does not, it will return [DeltaTableError::InvalidData] with a list
+    /// of values that violated each invariant.
+    pub fn check_batch(&self, record_batch: &RecordBatch) -> Result<(), DeltaTableError> {
+        self.enforce_invariants(record_batch)
+        // TODO: for support for Protocol V3, check constraints
+    }
+
+    fn enforce_invariants(&self, record_batch: &RecordBatch) -> Result<(), DeltaTableError> {
+        // Invariants are deprecated, so let's not pay the overhead for any of this
+        // if we can avoid it.
+        if self.invariants.is_empty() {
+            return Ok(());
+        }
+
+        let table = MemTable::try_new(record_batch.schema(), vec![vec![record_batch.clone()]])?;
+        self.ctx.register_table("data", Arc::new(table))?;
+
+        let mut violations: Vec<String> = Vec::new();
+
+        for invariant in self.invariants.iter() {
+            if invariant.field_name.contains('.') {
+                return Err(DeltaTableError::Generic(
+                    "Support for column invariants on nested columns is not supported.".to_string(),
+                ));
+            }
+
+            let sql = format!(
+                "SELECT {} FROM data WHERE not ({}) LIMIT 1",
+                invariant.field_name, invariant.invariant_sql
+            );
```

Comment:
how about using the dataframe api directly to skip the sql parsing overhead ;)

Reply:
I had started down that path, but I couldn't figure out how to pass a SQL string into the `.filter()` method. Any ideas on that?

Reply:
ha i see, you would need to call `sql_to_rex` to parse the expression, it's not worth it then in this case :D